File: src/core/intents.ts
import type { ShipIntent } from "../types";
import { state } from "./state";

// Cache: map of shipName -> set of target names claimed by that ship.
// Rebuilt on any intent change; one rebuild serves all callers per frame.
let _intentsDirty = true;
const _claimsByShip = new Map<string, Set<string>>();
const _intentChangeCallbacks: (() => void)[] = [];

/** Register a callback to be invoked whenever intents change (for cache invalidation). */
export function onIntentChange(cb: () => void): void {
	_intentChangeCallbacks.push(cb);
}

function rebuildClaimsCache(): void {
	_claimsByShip.clear();
	for (const [shipName, intent] of state.shipIntents) {
		const targets = new Set<string>();
		if (intent.type === "surveying") targets.add(intent.target);
		else if (intent.type === "transferring") targets.add(intent.destination);
		else if (intent.type === "tanking") targets.add(intent.target);
		if (targets.size > 0) _claimsByShip.set(shipName, targets);
	}
	_intentsDirty = false;
}

function markDirty(): void {
	_intentsDirty = true;
	for (const cb of _intentChangeCallbacks) cb();
}

/** Publish or update a ship's intent in the shared pool. */
export function publishIntent(shipName: string, intent: ShipIntent): void {
	state.shipIntents.set(shipName, intent);
	markDirty();
}

/** Remove a ship's intent from the pool (ship destroyed or removed). */
export function clearIntent(shipName: string): void {
	state.shipIntents.delete(shipName);
	markDirty();
}

/** Get the current intent for a specific ship, or undefined. */
export function getIntentForShip(shipName: string): ShipIntent | undefined {
	return state.shipIntents.get(shipName);
}

/** Check if a target body/asteroid is claimed by any ship OTHER than the given one. */
export function isTargetClaimed(targetName: string, excludeShipName: string): boolean {
	if (_intentsDirty) rebuildClaimsCache();
	for (const [shipName, targets] of _claimsByShip) {
		if (shipName === excludeShipName) continue;
		if (targets.has(targetName)) return true;
	}
	return false;
}

/** Force-invalidate the claimed targets cache (call after direct mutation of state.shipIntents in tests). */
export function invalidateIntentsCache(): void {
	_intentsDirty = true;
}

/** Get all target names claimed by ships other than the given one. Returns a Set for O(1) lookups. */
export function getClaimedTargets(excludeShipName: string): Set<string> {
	if (_intentsDirty) rebuildClaimsCache();
	const result = new Set<string>();
	for (const [shipName, targets] of _claimsByShip) {
		if (shipName === excludeShipName) continue;
		for (const t of targets) result.add(t);
	}
	return result;
}

File: rust-backend/crates/drift-sim/src/intents.rs
// Ship intent broadcast — mirrors src/core/intents.ts

use std::collections::HashSet;

use crate::state::{ShipIntent, State};

/// Publish (or overwrite) an intent for a ship.
pub fn publish_intent(ship_name: &str, intent: ShipIntent, state: &mut State) {
    state.ship_intents.insert(ship_name.to_string(), intent);
}

/// Remove the intent for a ship.
pub fn clear_intent(ship_name: &str, state: &mut State) {
    state.ship_intents.remove(ship_name);
}

/// Get the current intent for a ship, if any.
pub fn get_intent_for_ship<'a>(ship_name: &str, state: &'a State) -> Option<&'a ShipIntent> {
    state.ship_intents.get(ship_name)
}

/// Returns true if a target is claimed by any ship other than `exclude_ship`.
pub fn is_target_claimed(target: &str, exclude_ship: &str, state: &State) -> bool {
    for (ship_name, intent) in &state.ship_intents {
        if ship_name == exclude_ship {
            continue;
        }
        let claimed = match intent {
            ShipIntent::Surveying { target: t, .. } => t == target,
            ShipIntent::Transferring { destination: d, .. } => d == target,
            ShipIntent::SurveyPlan { targets, .. } => targets.iter().any(|t| t == target),
            _ => false,
        };
        if claimed {
            return true;
        }
    }
    false
}

/// Returns the set of targets claimed by all ships except `exclude_ship`.
pub fn get_claimed_targets(exclude_ship: &str, state: &State) -> HashSet<String> {
    let mut claimed = HashSet::new();
    for (ship_name, intent) in &state.ship_intents {
        if ship_name == exclude_ship {
            continue;
        }
        match intent {
            ShipIntent::Surveying { target, .. } => {
                claimed.insert(target.clone());
            }
            ShipIntent::Transferring { destination, .. } => {
                claimed.insert(destination.clone());
            }
            ShipIntent::SurveyPlan { targets, .. } => {
                for t in targets {
                    claimed.insert(t.clone());
                }
            }
            _ => {}
        }
    }
    claimed
}

/// Stub — in the full sim this invalidates a cached refuel-target.
/// Not needed for current tests but referenced from commands tests.
pub fn invalidate_intents_cache(_state: &mut State) {}

/// Returns true if any tanker ship has a Tanking intent targeting the given ship.
pub fn is_tanker_inbound_for(ship_name: &str, state: &State) -> bool {
    state
        .ship_intents
        .values()
        .any(|intent| matches!(intent, ShipIntent::Tanking { target, .. } if target == ship_name))
}
